File: 2022/day12/java/Solution.java
import java.io.*;
import java.nio.file.*;
import java.util.*;

public class Solution {
    private static char[][] grid;
    private static int rows, cols;
    private static int[] start = new int[2];
    private static int[] end = new int[2];

    public static void main(String[] args) throws IOException {
        String inputPath = Paths.get(System.getProperty("user.dir"), "..", "input.txt")
                                .normalize().toString();
        String text = Files.readString(Path.of(inputPath));
        parseGrid(text);

        System.out.println("Part 1: " + part1());
        System.out.println("Part 2: " + part2());
    }

    private static void parseGrid(String text) {
        String[] lines = text.trim().split("\n");
        rows = lines.length;
        cols = lines[0].length();
        grid = new char[rows][cols];

        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                char ch = lines[r].charAt(c);
                if (ch == 'S') {
                    start[0] = r;
                    start[1] = c;
                    grid[r][c] = 'a';
                } else if (ch == 'E') {
                    end[0] = r;
                    end[1] = c;
                    grid[r][c] = 'z';
                } else {
                    grid[r][c] = ch;
                }
            }
        }
    }

    private static int bfs(List<int[]> starts, int[] target) {
        boolean[][] visited = new boolean[rows][cols];
        LinkedList<int[]> queue = new LinkedList<>();

        for (int[] s : starts) {
            queue.add(new int[]{s[0], s[1], 0});
            visited[s[0]][s[1]] = true;
        }

        int[][] directions = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

        while (!queue.isEmpty()) {
            int[] current = queue.removeFirst();
            int r = current[0];
            int c = current[1];
            int dist = current[2];

            if (r == target[0] && c == target[1]) {
                return dist;
            }

            int currentHeight = grid[r][c];

            for (int[] dir : directions) {
                int nr = r + dir[0];
                int nc = c + dir[1];

                if (nr >= 0 && nr < rows && nc >= 0 && nc < cols && !visited[nr][nc]) {
                    int nextHeight = grid[nr][nc];
                    // Can move if destination is at most 1 higher
                    if (nextHeight <= currentHeight + 1) {
                        visited[nr][nc] = true;
                        queue.add(new int[]{nr, nc, dist + 1});
                    }
                }
            }
        }

        return -1; // No path found
    }

    private static int part1() {
        List<int[]> starts = new ArrayList<>();
        starts.add(start);
        return bfs(starts, end);
    }

    private static int part2() {
        // Find all cells with elevation 'a'
        List<int[]> starts = new ArrayList<>();
        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                if (grid[r][c] == 'a') {
                    starts.add(new int[]{r, c});
                }
            }
        }
        return bfs(starts, end);
    }
}

File: 2022/day12/go/solution.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"path/filepath"
)

type Point struct {
	row, col int
}

type State struct {
	row, col, dist int
}

func parseGrid(lines []string) ([][]byte, Point, Point) {
	grid := make([][]byte, len(lines))
	var start, end Point

	for r, line := range lines {
		grid[r] = []byte(line)
		for c, ch := range grid[r] {
			if ch == 'S' {
				start = Point{r, c}
				grid[r][c] = 'a'
			} else if ch == 'E' {
				end = Point{r, c}
				grid[r][c] = 'z'
			}
		}
	}

	return grid, start, end
}

func bfs(grid [][]byte, starts []Point, end Point) int {
	rows := len(grid)
	cols := len(grid[0])
	visited := make(map[Point]bool)
	queue := make([]State, 0)

	for _, start := range starts {
		queue = append(queue, State{start.row, start.col, 0})
		visited[start] = true
	}

	directions := []Point{{-1, 0}, {1, 0}, {0, -1}, {0, 1}}

	for len(queue) > 0 {
		current := queue[0]
		queue = queue[1:]

		if current.row == end.row && current.col == end.col {
			return current.dist
		}

		currentHeight := grid[current.row][current.col]

		for _, d := range directions {
			nr, nc := current.row+d.row, current.col+d.col

			if nr >= 0 && nr < rows && nc >= 0 && nc < cols {
				nextPoint := Point{nr, nc}
				if !visited[nextPoint] {
					nextHeight := grid[nr][nc]
					// Can move if destination is at most 1 higher
					if nextHeight <= currentHeight+1 {
						visited[nextPoint] = true
						queue = append(queue, State{nr, nc, current.dist + 1})
					}
				}
			}
		}
	}

	return -1 // No path found
}

func part1(lines []string) int {
	grid, start, end := parseGrid(lines)
	return bfs(grid, []Point{start}, end)
}

func part2(lines []string) int {
	grid, _, end := parseGrid(lines)

	// Find all cells with elevation 'a'
	var starts []Point
	for r, row := range grid {
		for c, ch := range row {
			if ch == 'a' {
				starts = append(starts, Point{r, c})
			}
		}
	}

	return bfs(grid, starts, end)
}

func main() {
	// Get the directory of the executable
	execPath, _ := os.Executable()
	execDir := filepath.Dir(execPath)

	// Try multiple paths for input file
	inputPaths := []string{
		filepath.Join(execDir, "..", "input.txt"),
		"../input.txt",
	}

	var file *os.File
	var err error

	for _, path := range inputPaths {
		file, err = os.Open(path)
		if err == nil {
			break
		}
	}

	if err != nil {
		fmt.Fprintf(os.Stderr, "Error opening input file: %v\n", err)
		os.Exit(1)
	}
	defer file.Close()

	var lines []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		if line != "" {
			lines = append(lines, line)
		}
	}

	if err := scanner.Err(); err != nil {
		fmt.Fprintf(os.Stderr, "Error reading input: %v\n", err)
		os.Exit(1)
	}

	fmt.Println("Part 1:", part1(lines))
	fmt.Println("Part 2:", part2(lines))
}
